File: planning-agent/app/transport/events.py
"""Event envelope shared between Planning Agent and the orchestrator.

Mirrors `orchestrator/src/websocket/events.ts` — same field names, same
event types. Python uses snake_case natively, TypeScript was authored to
match, so the wire representation is identical.
"""
from __future__ import annotations

import time
from typing import Any, Literal, TypedDict

AgentSource = Literal["planning", "building"]

AgentEventType = Literal[
    "progress",
    "log",
    "error",
    "user_prompt",
    "phase_start",
    "phase_end",
    "token",
    "tool_call",
    "tool_result",
    "completion",
]


class AgentEvent(TypedDict, total=False):
    agent: AgentSource
    project_id: str
    session_id: str
    build_id: str
    event_type: AgentEventType
    phase: str
    progress_percent: int
    payload: Any
    at: int


def make_event(
    event_type: AgentEventType,
    project_id: str,
    *,
    session_id: str | None = None,
    phase: str | None = None,
    progress_percent: int | None = None,
    payload: Any = None,
) -> AgentEvent:
    event: AgentEvent = {
        "agent": "planning",
        "project_id": project_id,
        "event_type": event_type,
        "at": int(time.time() * 1000),
    }
    if session_id:
        event["session_id"] = session_id
    if phase:
        event["phase"] = phase
    if progress_percent is not None:
        event["progress_percent"] = progress_percent
    if payload is not None:
        event["payload"] = payload
    return event

File: orchestrator/src/websocket/events.ts
/**
 * Unified event schema for agent ↔ orchestrator ↔ frontend communication.
 * See ARCHITECTURE.md §12.3.
 *
 * Every event carries enough context that the frontend can route it to the
 * right project view without extra lookups.
 */

export type AgentSource = 'planning' | 'building';

export type AgentEventType =
  | 'progress'       // phase advance, percent update
  | 'log'            // free-text log line (human-readable)
  | 'error'          // unrecoverable error within the agent
  | 'user_prompt'    // agent-initiated question to the user (E1)
  | 'phase_start'    // building: new phase begins
  | 'phase_end'      // building: phase concludes (success/fail)
  | 'token'          // planning: streaming assistant token
  | 'tool_call'      // planning: tool invocation (name + args)
  | 'tool_result'    // planning: tool returned
  | 'completion';    // turn/run finished

export interface AgentEvent<TPayload = unknown> {
  agent: AgentSource;
  project_id: string;
  /** Optional — only populated once a session exists. */
  session_id?: string;
  /** Optional — only populated during build runs. */
  build_id?: string;
  event_type: AgentEventType;
  /** Optional — phase name for build events, conversation phase for planning. */
  phase?: string;
  /** 0 - 100. Not all events carry progress (e.g. tokens). */
  progress_percent?: number;
  payload?: TPayload;
  /** UNIX ms; populated by the gateway on emit. */
  at?: number;
}
